doc/source/file_formats.rst

```diff
@@ -1,23 +1,375 @@
+as its first content line. This is followed by one or more
+subsections. The subsection atoms is mandatory, optional subsections
+are edges and subsections corresponding to the different types of
+atomic interactions, including bonds, constraints, angles, and
+dihedrals.
+
+Subsection metadata
+^^^^^^^^^^^^^^^^^^^
+
+Subsections can be given metadata, using a #meta directive. These
+directives always apply to the whole subsection. This makes it
+possible to, e.g., group interactions or annotate bonds according to
+their type. It is also possible to specify a context for the
+interactions in a subsection, using 'ifdef' and/or 'ifndef' tags. This
+will make the inclusion of the interactions conditional by adding
+#ifdef and #ifndef directive statements to the output topology
+files. The metadata is given as a JSON style mapping of key/value
+pairs.
+
+#meta {"group": "Side chain bonds", "ifdef": "FLEXIBLE"}
+
+Metadata can also be added to a single line by adding an attribute
+statement as the last element.
+
+Atoms
+^^^^^
+
+Each line in the atoms section describes one particle, corresponding
+to a node in the molecular graph. The description comprises the atom
+name, atom type, residue identifier, mass and charge.
+
+Edges
+^^^^^
+
+Edges will be added to the molecular graph when required through
+interactions, but they can also be added explicitly by listing them
+under the edges subsections, specifying each edge to be added by the
+corresponding atom names.
+
+Interactions
+^^^^^^^^^^^^
+
+There are several options for subsections describing interactions
+between particles. Of these, bonds, angles, dihedrals, cmap, and
+constraints will automatically add the corresponding edges to the
+molecular graph, unless specified explicitly by setting an attribute
+'edge' to false in a subsection #meta or following a specific
+interaction.
+
+Each line in an interactions subsection specifies one interaction by
+listing the atoms involved by name, followed by the interaction
+parameters. For all interactions, the parameters are read as is and
+written to the output topology without interpreting and/or
+checking. Bond/constraint lengths, angles and dihedral angles may be
+used for generating missing coordinates.
+
+A full list of interactions is given below, corresponding to the list
+of intramolecular interactions available in Gromacs, with a number
+specifying the number of particles involved in the interaction. Note
+that improper dihedrals are listed as a separate interaction type,
+whereas in Gromacs these fall under the dihedrals section.
+
+* bonds(2)
+* angles(3)
+* dihedrals(4)
+* impropers(4)
+* constraints(2)
+* pairs(2)
+* pairs_nb(2)
+* SETTLE(1)
+* virtual_sites2(3)
+* virtual_sites3(4)
+* virtual_sites4(5)
+* position_restraints(1)
+* distance_restraints(2)
+* dihedral_restraints(4)
+* orientation_restraints(2)
+* angle_restraints(4)
+* angle_restraints_z(2)
+* cmap(...)
+
```

Comment:
virtual_sitesn needs special mention, since it takes an unknown number of atoms, and the syntax is different (is it still?) from the gromacs syntax:
gmx: `<at_id1> <func> <at_ids ...>`
M2: `<at_id1> <at_ids...> -- <func>`

Why is this? So that we can assume (only) the first X elements are atom names. Everything after `--` is parameters. This was added to be able to kind-of support unknown interactions, but that functionality perished later on. This `--` also works for other interactions. If you insist on making a bond with 3 atoms:
`at1 at2 at3 -- 2 0.1 400`